File: iguagile/room.go
package iguagile

import (
	"context"
	"errors"
	"fmt"
	"log"
	"math"
	"os"

	pb "github.com/iguagile/iguagile-room-proto/room"
)

// Room maintains the set of active clients and broadcasts messages to the
// clients.
type Room struct {
	clientManager    *ClientManager
	generator        *idGenerator
	log              *log.Logger
	host             *Client
	config           *RoomConfig
	creatorConnected bool
	roomProto        *pb.Room
	store            Store
	engine           *Engine
	service          RoomService
	streams          map[string]*Stream
	ready            bool
}

// RoomConfig is room config.
type RoomConfig struct {
	RoomID          int
	ApplicationName string
	Version         string
	Password        string
	MaxUser         int
	Info            map[string]string
	Token           []byte
}

func newRoom(engine *Engine, config *RoomConfig) (*Room, error) {
	gen, err := newIDGenerator()
	if err != nil {
		return nil, err
	}

	return &Room{
		clientManager: NewClientManager(),
		generator:     gen,
		log:           log.New(os.Stdout, "iguagile-engine ", log.Lshortfile),
		config:        config,
		store:         engine.store,
		roomProto:     &pb.Room{},
		engine:        engine,
		streams:       map[string]*Stream{},
	}, nil
}

func (r *Room) serve(ctx context.Context, conn *quicConn) error {
	client, err := NewClient(r, conn)
	if err != nil {
		return err
	}

	r.roomProto.ConnectedUser = int32(r.clientManager.count + 1)
	if err := r.store.RegisterRoom(r.roomProto); err != nil {
		r.log.Println(err)
	}
	return r.register(ctx, client)
}

const (
	// Maximum message size allowed from peer.
	maxMessageSize = math.MaxUint16
)

// register requests from the clients.
func (r *Room) register(ctx context.Context, client *Client) error {
	if err := r.clientManager.Add(client); err != nil {
		return err
	}

	if r.clientManager.Count() == 1 {
		r.host = client
	}

	for name, s := range r.streams {
		stream, err := client.conn.OpenStream()
		if err != nil {
			return err
		}

		if _, err := stream.Write([]byte(name)); err != nil {
			return err
		}

		s.streams[client.id] = stream
		client.streams[name] = stream
	}

	client.readStart(ctx)

	return r.service.OnRegisterClient(client.id)
}

// unregister requests from clients.
func (r *Room) unregister(client *Client) error {
	if err := r.generator.free(client.GetID()); err != nil {
		r.log.Println(err)
	}

	r.clientManager.Remove(client.GetID())
	if client == r.host {
		c, err := r.clientManager.First()
		if err != nil {
			return err
		}
		r.host = c
	}

	return r.service.OnUnregisterClient(client.id)
}

// CreateStream creates new stream.
// Call in the Create method implemented in RoomServiceFactory.
func (r *Room) CreateStream(streamName string) (*Stream, error) {
	if r.ready {
		return nil, errors.New("call CreateStream in the Create method implemented in RoomServiceFactory")
	}

	streamName = userStreamPrefix + streamName

	if _, ok := r.streams[streamName]; ok {
		return nil, fmt.Errorf("%v has already been created", streamName)
	}

	stream := &Stream{
		r:       r,
		streams: map[int]*quicStream{},
	}

	r.streams[streamName] = stream

	return stream, nil
}

// CloseConnection closes the connection and unregisters the client.
func (r *Room) CloseConnection(client *Client) {
	if err := r.unregister(client); err != nil {
		r.log.Println(err)
	}
	if err := client.Close(); err != nil && err.Error() != "use of closed network connection" {
		r.log.Println(err)
	}
}

// Close closes all client connections.
func (r *Room) Close() error {
	r.clientManager.Lock()
	defer r.clientManager.Unlock()
	for _, client := range r.clientManager.GetAllClients() {
		if err := client.Close(); err != nil {
			r.log.Println(err)
		}
	}

	r.engine.rooms.Delete(r.config.RoomID)
	return r.service.Destroy()
}

File: src/room.rs
use std::collections::HashMap;
use std::rc::Rc;

use anyhow::{anyhow, Error};

use crate::client::{Client, QUICClient};
use crate::engine::Engine;
use crate::id::{IdPool, MemoryIdPool};
use crate::store::{iguagile, MemoryStore, Store};

pub struct RoomConfig {
    room_id: u16,
    application_name: String,
    version: String,
    password: String,
    max_user: u16,
    info: HashMap<String, String>,
    token: Vec<u8>,
}

trait RoomTrait {
    fn join(&mut self, c: Box<dyn Client>) -> Result<(), anyhow::Error>;
    fn leave(&mut self, c: Box<dyn Client>) -> Result<(), anyhow::Error>;
}

trait ClientManagerTrait {
    fn get(&self, client_id: u16) -> Result<&dyn Client, anyhow::Error>;
    fn add(&mut self, client_id: u16, c: &dyn Client) -> Result<(), anyhow::Error>;
    fn remove(&mut self, client_id: u16) -> Result<(), anyhow::Error>;
    fn exists(&self, client_id: u16) -> bool;
    fn count(&self) -> u16;
    fn clear(&mut self);
    fn first(&self) -> Result<&dyn Client, anyhow::Error>;
}

struct ClientManager {
    clients: Rc<HashMap<u16, Box<dyn Client>>>,
}

impl ClientManager {
    pub fn new() -> Self {
        ClientManager {
            clients: Rc::new(HashMap::new()),
        }
    }
}

impl ClientManagerTrait for ClientManager {
    fn get(&self, client_id: u16) -> Result<&dyn Client, Error> {
        todo!()
    }

    fn add(&mut self, client_id: u16, c: &dyn Client) -> Result<(), Error> {
        todo!()
    }

    fn remove(&mut self, client_id: u16) -> Result<(), Error> {
        todo!()
    }

    fn exists(&self, client_id: u16) -> bool {
        todo!()
    }

    fn count(&self) -> u16 {
        todo!()
    }

    fn clear(&mut self) {
        todo!()
    }

    fn first(&self) -> Result<&dyn Client, Error> {
        todo!()
    }
}

pub struct Room {
    clients: Box<dyn ClientManagerTrait>,
    id_pool: Box<dyn IdPool>,
    host: Box<dyn Client>,
    config: RoomConfig,
    creator_connected: bool,
    room_proto: iguagile::Room,
    store: Box<dyn Store>,
    engine: Engine,
    // service: RoomService,
    streams: HashMap<String, Rc<dyn Client>>,
    ready: bool,
}

impl Room {
    pub fn new(config: RoomConfig) -> Self {
        Room {
            clients: Box::new(ClientManager::new()),
            id_pool: Box::new(MemoryIdPool::new()),
            host: Box::new(QUICClient::new(1, "127.0.0.1").unwrap()),
            config: config,
            creator_connected: false,
            room_proto: iguagile::Room::default(),
            store: Box::new(MemoryStore::new(redis::Client::open("redis://").unwrap())),
            engine: Engine::new(
                1,
                Box::new(MemoryStore::new(redis::Client::open("redis://").unwrap())),
                Box::new(MemoryIdPool::new()),
                &iguagile::Server::default(),
                std::time::Duration::from_secs(1),
                std::time::Duration::from_secs(1),
            ),
            streams: HashMap::new(),
            ready: false,
        }
    }
}

impl RoomTrait for Room {
    fn join(&mut self, c: Box<dyn Client>) -> Result<(), anyhow::Error> {
        let id = c.get_id();
        if self.clients.exists(id) {
            return Err(anyhow::anyhow!("client already joined"));
        }
        self.clients.add(id, &*c)
    }

    fn leave(&mut self, c: Box<dyn Client>) -> Result<(), anyhow::Error> {
        let id = c.get_id();
        if !self.clients.exists(id) {
            return Err(anyhow::anyhow!("client not joined"));
        }
        self.clients.remove(id)
    }
}
